src/libponyc/ast/parser.c

```diff
@@ -118,10 +118,27 @@ DEF(typearg);
   RULE("type argument", type, typeargliteral, typeargconst);
   DONE();
 
-// ID [COLON type] [ASSIGN typearg]
+// UNDERSCORE
+DEF(underscore);
+    AST_NODE(TK_UNBOUNDTYPEPARAM);
+    TOKEN("_", TK_ID);
+    IF(TK_COLON, RULE("type constraint", type));
+    IF(TK_ASSIGN, RULE("default type argument", typearg));
+    DONE();
+
+// LSQUARE UNDERSCORE {COMMA UNDERSCORE} RSQUARE
+DEF(unboundtypeargs);
```

Comment:
I think you want this to have an `AST_NODE` macro at the top of it, so that the list of underscores has somewhere to land. This may be the reason why you're having trouble with the `OPT` macro not working as expected?

Also, just a question, but if this node is a list of `TK_UNBOUNDTYPEPARAM`s, then shouldn't it be called `unboundtypeparams` rather than `unboundtypeargs` (as it is currently called)?

Reply:
I added an `AST_NODE` and created `TK_UNBOUNDEDTYPEPARAMS` which is the optional sibling of the typeparam `ID`. Unfortunately, I'm still running into `p != NULL` issues. Still debugging.